### PR TITLE
Partial fix for 6674. 

### PR DESCRIPTION
**Pull request by lajones**
Partial fix for #6674

Created new Navigation() calls directly on EntityBuilder. Added tests for OneToMany, ManyToOne, ManyToMany, OneToOne and principal side of the Owned Type.

#### Things remaining:
- The additions to `WithOwner()` to allow configuring a navigation on the dependent side of an Owned Type.
- Removing the old way of doing it (see #20195).

### CODE REVIEW

File: test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs [hunk 1]
```diff
@@ -1329,6 +1329,80 @@ public virtual void Inheritance_where_base_has_multiple_owned_types_works()
 
                 Assert.Equal(4, model.GetEntityTypes().Count());
             }
+
+            [ConditionalFact]
+            public virtual void Principal_end_of_navigations_to_OneToOne_Owned_types_can_set_access_mode_using_expressions()
```
Comment (lajones) [inline test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs]: Should rename OneToOne -> OwnsOne

File: test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs [hunk 2]
```diff
@@ -1329,6 +1329,80 @@ public virtual void Inheritance_where_base_has_multiple_owned_types_works()
...
+                    .Navigation("OwnedDependent")
+                    .UsePropertyAccessMode(PropertyAccessMode.Field);
+
+                var principal = (IEntityType)model.FindEntityType(typeof(OneToOneNavPrincipalOwner));
+
+                Assert.Equal(PropertyAccessMode.Field, principal.FindNavigation("OwnedDependent").GetPropertyAccessMode());
+            }
+
+            [ConditionalFact]
+            public virtual void Principal_end_of_navigations_to_OneToMany_Owned_types_can_set_access_mode_using_expressions()
```
Comment (lajones) [inline test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs]: Should rename OneToMany -> OwnsMany

File: src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs [hunk 3]
```diff
@@ -127,7 +127,7 @@ public class CollectionNavigationBuilder : IInfrastructure<InternalRelationshipB
         /// <returns> An object to further configure the relationship. </returns>
         public virtual ReferenceCollectionBuilder WithOne(
             [CanBeNull] string navigationName = null,
-            [CanBeNull] Action<NavigationBuilder> navigationConfiguration = null)
+            [CanBeNull] Action<NavigationIdentityBuilder> navigationConfiguration = null)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs]: Why did you rename `NavigationBuilder`?

File: src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    /// <summary>
+    ///     <para>
+    ///         Provides a simple API for configuring a <see cref="ISkipNavigation" />.
+    ///     </para>
+    ///     <para>
+    ///         Instances of this class are returned from methods when using the <see cref="ModelBuilder" /> API
+    ///         and it is not designed to be directly constructed in your application code.
+    ///     </para>
+    /// </summary>
+    public class SkipNavigationBuilder : IInfrastructure<IConventionSkipNavigationBuilder>
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs]: Why is this class needed?

File: src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace Microsoft.EntityFrameworkCore.Metadata.Internal
+{
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public class InternalNavigationIdentityBuilder
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs]: Why is this class needed? This can be handled in the public builder

File: src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs (hunk 4 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs]: ManyToMany can result in there being `SkipNavigation`s instead of `Navigation`s.

File: src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs (hunk 3 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs]: `NavigationBuilder` is still there. `NavigationIdentityBuilder` is the class that works regardless of whether you're using `Navigation`s or `SkipNavigation`s. Works like `MemberIdentity`.

File: src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs (hunk 5 above)
Comment (lajones) [reply src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs]: See replies above.

**Comment by lajones** [comment]
@AndriySvyryd See also #20276 for being able to configure the dependent side navigation of an Owned-type.

File: src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs (hunk 5 above)
Comment (lajones) [reply src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs]: `InternalEntityTypeBuilder` has references to `InternalNavigationIdentityBuilder` for its `Navigation()` methods (just as the `Property()` methods have `InternalPropertyBuilder`).

It has to be `InternalNavigationIdentityBuilder` and not `InternalNavigationBuilder` for the same reason as above - sometimes it's `SkipNavigation`s we're processing.

File: src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs]: Right, but the new `NavigationIdentityBuilder` handles those already

File: src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs]: I just don't see why you would need to create a new class instead of making these changes to the existing `NavigationBuilder`

File: src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs (hunk 5 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs]: This can be handled in the new `NavigationIdentityBuilder`, we don't need an additional class

Comment (lajones) [reply src/EFCore/Metadata/Internal/InternalNavigationIdentityBuilder.cs]: Hmmm. Every other call in `InternalEntityTypeBuilder` looks like this. It seems very odd to be mixing the Internal and non-Internal parts. Can we discuss?

File: src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs (hunk 4 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs]: Hmm. It handles it by passing it to the `SkipNavigationBuilder`.

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/SkipNavigationBuilder.cs]: But it doesn't need to, it can call to `InternalSkipNavigationBuilder` directly

**Comment by AndriySvyryd** [comment]
I pushed a commit for what I was talking about

**Comment by lajones** [comment]
@AndriySvyryd Ah - understand now. Thanks.

File: src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs [hunk 6]
```diff
@@ -134,6 +134,20 @@ public virtual PropertyBuilder<TProperty> Property<TProperty>([NotNull] Expressi
...
+        /// <summary>
+        ///         Returns an object that can be used to configure an existing navigation property of the entity type.
+        ///         It is an error for the navigation property not to exist.
+        /// </summary>
+        /// <param name="navigationExpression">
+        ///     A lambda expression representing the navigation property to be configured (
+        ///     <c>blog => blog.Posts</c>).
+        /// </param>
+        /// <returns> An object that can be used to configure the navigation property. </returns>
+        public virtual NavigationBuilder Navigation<TNavigation>([NotNull] Expression<Func<TEntity, TNavigation>> navigationExpression)
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs]: Do we need generic navigationBuilder?

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs [hunk 7]
```diff
@@ -27,22 +27,25 @@ public class NavigationBuilder : IInfrastructure<IConventionNavigationBuilder>
...
         /// </summary>
         [EntityFrameworkInternal]
-        public NavigationBuilder([NotNull] IMutableNavigation navigation)
+        public NavigationBuilder([NotNull] IMutableNavigationBase navigationOrSkipNavigation)
         {
-            Check.NotNull(navigation, nameof(navigation));
+            Check.NotNull(navigationOrSkipNavigation, nameof(navigationOrSkipNavigation));
 
-            Builder = ((Navigation)navigation).Builder;
+            NavBuilder = (navigationOrSkipNavigation as Navigation)?.Builder;
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/NavigationBuilder.cs]: NavigationBuilder.

Comment (smitpatel) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: Does it has to be concrete type Navigation?

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs [hunk 8]
```diff
@@ -27,22 +27,25 @@ public class NavigationBuilder : IInfrastructure<IConventionNavigationBuilder>
...
+        private InternalNavigationBuilder NavBuilder { get; }
+
+        private InternalSkipNavigationBuilder SkipNavBuilder { get; }
 
         /// <summary>
         ///     The navigation being configured.
         /// </summary>
-        public virtual IMutableNavigation Metadata => Builder.Metadata;
+        public virtual IMutableNavigationBase Metadata
+            => (IMutableNavigationBase)NavBuilder?.Metadata ?? SkipNavBuilder?.Metadata;
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/NavigationBuilder.cs]: Better to store than recompute?

File: src/EFCore/Properties/CoreStrings.resx [hunk 9]
```diff
@@ -1357,4 +1357,7 @@
   <data name="AttemptToCreateEntityTypeBasedOnProxyClass" xml:space="preserve">
     <value>Cannot add an entity type with type '{typeName}'. That type is a dynamically-generated proxy type.</value>
   </data>
+  <data name="CanOnlyConfigureExistingNavigations" xml:space="preserve">
+    <value>There is no navigation property with name '{navigationName}' on entity type '{entityType}'. Only pre-existing navigations can be configured.</value>
```
Comment (smitpatel) [inline src/EFCore/Properties/CoreStrings.resx]: pre-existing is really bad term here.

File: test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs]: Rename it, it is test only.

File: src/EFCore/Properties/CoreStrings.resx (hunk 9 above)
Comment (lajones) [reply src/EFCore/Properties/CoreStrings.resx]: Please suggest alternatives. I don't want to get into "that's no good either" conversations.

Comment (smitpatel) [reply src/EFCore/Properties/CoreStrings.resx]: Please add the navigation to model before configuring it.

**Comment by smitpatel** [comment]
> Removing the old way of doing it (see #20195).

Not in this PR. Where are those changes?

**Comment by lajones** [comment]
@smitpatel I'm not removing the old way until we're agreed this is the way forward.

File: src/EFCore/Properties/CoreStrings.resx (hunk 9 above)
Comment (lajones) [reply src/EFCore/Properties/CoreStrings.resx]: Sounds good.

**Comment by lajones** [comment]
@smitpatel The tests are renamed in the other PR (#20276) - where we allow you to specify the dependent end as well.

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs (hunk 7 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: Yes. `INavigation` and `ISkipNavigation` do not have `Builder` defined.

File: src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs (hunk 6 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs]: Would probably be better. @AndriySvyryd would you mind the extra class?

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs]: Not yet

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs]: We'll need it for `.HasField()`

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs (hunk 7 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: NavigationBuilder clashes with class name.
We need the internal builder

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs (hunk 8 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: Will do

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: This follows the pattern in other builders

**Comment by AndriySvyryd** [comment]
Squash the commits

**Comment by lajones** [comment]
@Andriy Need to change the names in the tests to get rid of NavigationIdentityBuilder there too. Then will do.

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs [hunk 10]
```diff
@@ -27,22 +27,25 @@ public class NavigationBuilder : IInfrastructure<IConventionNavigationBuilder>
...
-        public NavigationBuilder([NotNull] IMutableNavigation navigation)
+        public NavigationBuilder([NotNull] IMutableNavigationBase navigationOrSkipNavigation)
         {
-            Check.NotNull(navigation, nameof(navigation));
+            Check.NotNull(navigationOrSkipNavigation, nameof(navigationOrSkipNavigation));
 
-            Builder = ((Navigation)navigation).Builder;
+            NavBuilder = (navigationOrSkipNavigation as Navigation)?.Builder;
+            SkipNavBuilder = (navigationOrSkipNavigation as SkipNavigation)?.Builder;
+            Metadata = (IMutableNavigationBase)NavBuilder?.Metadata ?? SkipNavBuilder?.Metadata;
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/NavigationBuilder.cs]: The `Metadata` property should be computed, it's rarely accessed.
If you are going to store it, then assign from `navigationOrSkipNavigation`

File: src/EFCore/Properties/CoreStrings.resx [hunk 11]
```diff
@@ -1357,4 +1357,7 @@
   <data name="AttemptToCreateEntityTypeBasedOnProxyClass" xml:space="preserve">
     <value>Cannot add an entity type with type '{typeName}'. That type is a dynamically-generated proxy type.</value>
   </data>
+  <data name="CanOnlyConfigureExistingNavigations" xml:space="preserve">
+    <value>There is no navigation property with name '{navigationName}' on entity type '{entityType}'. Please add the navigation to the model before configuring it.</value>
```
Comment (AndriySvyryd) [inline src/EFCore/Properties/CoreStrings.resx]:
```suggestion
    <value>There is no navigation property with name '{navigationName}' on entity type '{entityType}'. Please add the navigation to the entity type before configuring it.</value>
```

File: test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs (hunk 2 above)
Comment (lajones) [reply test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs]: I'll address in PR 20276 - I'm renaming them there anyway - as it will no longer be just the principal side being tested.

File: test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs [hunk 12]
```diff
@@ -158,6 +159,56 @@ public virtual void Throws_for_conflicting_many_to_one_on_right()
...
+
+            [ConditionalFact]
+            public virtual void Navigation_properties_can_set_access_mode_using_navigation_names()
+            {
+                var modelBuilder = CreateModelBuilder();
+                var model = modelBuilder.Model;
+
+                modelBuilder.Entity<ManyToManyNavPrincipal>()
+                    .HasMany<NavDependent>("Dependents")
+                    .WithMany("ManyToManyPrincipals");
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs]: You should use names that are different from the CLR properties to get better coverage.

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs (hunk 8 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: @AndriySvyryd - any objection - storing it in this PR.

File: src/EFCore/Properties/CoreStrings.resx (hunk 11 above)
Comment (lajones) [reply src/EFCore/Properties/CoreStrings.resx]: Will do.

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs (hunk 10 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: You mean `(navigationOrSkipNavigation as Navigation)?.Builder.Metadata ?? (navigationOrSkipNavigation as SkipNavigation)?.Builder.Metadata`? That seems worse?

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: `Metadata = navigationOrSkipNavigation`

File: test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs (hunk 12 above)
Comment (lajones) [reply test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs]: I tried what I thought you meant and I get _The navigation property 'NonClrDependents' cannot be added to the entity type 'ManyToManyNavPrincipal' because there is no corresponding CLR property on the underlying type and navigations properties cannot be added to shadow state._
Can you explain more?

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs (hunk 10 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: Doh!

File: test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs (hunk 12 above)
Comment (AndriySvyryd) [reply test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs]: Oh, you also need to use shadow entity types

File: src/EFCore/Metadata/Builders/NavigationBuilder.cs (hunk 10 above)
Comment (lajones) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: Just FYI why don't we do this on all of the builders?

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: It uses more memory and is completely unnecessary

Comment (lajones) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: Do you want to just go back to doing it your way then? It looked good to me but I'm happy to revert.

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/NavigationBuilder.cs]: I'd prefer that, but I don't feel strongly

File: test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs (hunk 12 above)
Comment (lajones) [reply test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs]: Let's add that separately. It sounds likes that going to be tricky given the error above "_cannot be added to shadow state_". I'll add it to the issue to make sure I don't forget. I want to get feedback that the rest of the team likes this new approach before I flesh out the tests.
